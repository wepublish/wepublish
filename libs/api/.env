DATABASE_URL=postgresql://postgres@localhost:5432/wepublish_test?schema=public