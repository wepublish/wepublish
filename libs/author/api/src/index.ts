export * from './lib/author.module'
export * from './lib/author.resolver'
export * from './lib/author.service'
