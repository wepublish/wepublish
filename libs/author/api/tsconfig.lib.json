{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "../../../dist/out-tsc",
    "declaration": true,
    "types": ["node"],
    "target": "es2021",
    "strictNullChecks": true,
    "noImplicitAny": true,
    "strictBindCallApply": true,
    "forceConsistentCasingInFileNames": true,
    "noFallthroughCasesInSwitch": true
  },
  "include": ["src/**/*.ts", "../../article/api/src/lib/author.service.ts"],
  "exclude": ["jest.config.ts", "src/**/*.spec.ts", "src/**/*.test.ts"]
}
