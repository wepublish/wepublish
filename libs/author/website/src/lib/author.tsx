import {styled} from '@mui/material'
import {
  BuilderAuthorLinksProps,
  BuilderAuthorProps,
  useWebsiteBuilder
} from '@wepublish/website/builder'

declare module 'react' {
  interface HTMLAttributes<T> {
    fetchPriority?: 'high' | 'low' | 'auto'
  }
}

export const AuthorWrapper = styled('article')`
  display: grid;
  gap: ${({theme}) => theme.spacing(3)};
`

export function Author({
  className,
  data,
  authorLinks: AuthorLinksComponent = AuthorLinks,
  loading,
  error
}: BuilderAuthorProps) {
  const {
    elements: {Image, H3, H5},
    blocks: {RichText}
  } = useWebsiteBuilder()

  if (!data?.author) {
    return null
  }

  return (
    <AuthorWrapper className={className}>
      <header>
        <H3 component="h1">{data.author.name}</H3>
        {data.author.jobTitle && <H5 component="aside">{data.author.jobTitle}</H5>}
      </header>

      {data.author.image && <Image image={data.author.image} fetchPriority="high" />}
      {!!data.author.links?.length && <AuthorLinksComponent links={data.author.links} />}

      <RichText richText={data.author.bio ?? []} />
    </AuthorWrapper>
  )
}

const AuthorLinksWrapper = styled('aside')`
  display: grid;
  grid-auto-columns: max-content;
  grid-auto-flow: column;
  gap: ${({theme}) => theme.spacing(2)};
`

export function AuthorLinks({links}: BuilderAuthorLinksProps) {
  const {
    elements: {Link}
  } = useWebsiteBuilder()

  return (
    <AuthorLinksWrapper>
      {links.map((link, index) => (
        <Link key={index} href={link.url} target="__blank">
          {link.title}
        </Link>
      ))}
    </AuthorLinksWrapper>
  )
}
