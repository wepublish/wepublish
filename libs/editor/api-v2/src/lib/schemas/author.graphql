fragment FullAuthor on Author {
  id
  name
  jobTitle
  slug
  bio
  url
  createdAt
  modifiedAt

  links {
    title
    url
  }

  tags {
    ...FullTag
  }

  image {
    ...FullImage
  }
  hideOnArticle
  hideOnTeaser
  hideOnTeam
}
