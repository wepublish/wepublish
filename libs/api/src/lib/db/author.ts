import {RichTextNode} from '@wepublish/richtext/api'

export interface AuthorLink {
  title: string
  url: string
}

export interface Author {
  id: string
  createdAt: Date
  modifiedAt: Date
  name: string
  slug: string
  jobTitle?: string | null
  imageID?: string | null
  links: AuthorLink[]
  bio: RichTextNode[]
}

export interface AuthorInput {
  name: string
  slug: string
  jobTitle?: string | null
  imageID?: string | null
  links: AuthorLink[]
  bio: RichTextNode[]
}

export type OptionalAuthor = Author | null

export enum AuthorSort {
  CreatedAt = 'createdAt',
  ModifiedAt = 'modifiedAt',
  Name = 'name'
}

export interface AuthorFilter {
  name?: string
}
