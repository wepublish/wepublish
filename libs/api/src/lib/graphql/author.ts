import {
  GraphQLNonNull,
  GraphQLID,
  GraphQLString,
  GraphQLObjectType,
  GraphQLList,
  GraphQLInt,
  GraphQLInputObjectType,
  GraphQLEnumType
} from 'graphql'

import {Author, AuthorSort} from '../db/author'
import {Context} from '../context'

import {GraphQLPageInfo} from './common'
import {GraphQLImage} from './image'
import {GraphQLSlug} from './slug'
import {GraphQLRichText} from '@wepublish/richtext/api'
import {GraphQLDateTime} from 'graphql-scalars'
import {createProxyingResolver} from '../utility'
import {GraphQLTag} from './tag/tag'

export const GraphQLAuthorLink = new GraphQLObjectType<Author, Context>({
  name: 'AuthorLink',
  fields: {
    title: {type: new GraphQLNonNull(GraphQLString)},
    url: {type: new GraphQLNonNull(GraphQLString)}
  }
})

export const GraphQLAuthor = new GraphQLObjectType<Author, Context>({
  name: 'Author',
  fields: {
    id: {type: new GraphQLNonNull(GraphQLID)},

    createdAt: {type: new GraphQLNonNull(GraphQLDateTime)},
    modifiedAt: {type: new GraphQLNonNull(GraphQLDateTime)},

    name: {type: new GraphQLNonNull(GraphQLString)},
    slug: {type: new GraphQLNonNull(GraphQLSlug)},
    url: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: createProxyingResolver(async (author, _, {urlAdapter}) => {
        return await urlAdapter.getAuthorURL(author)
      })
    },
    links: {type: new GraphQLList(new GraphQLNonNull(GraphQLAuthorLink))},
    bio: {type: GraphQLRichText},
    jobTitle: {type: GraphQLString},
    image: {
      type: GraphQLImage,
      resolve: createProxyingResolver(({imageID}, args, {loaders}) => {
        return imageID ? loaders.images.load(imageID) : null
      })
    },
    tags: {
      type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(GraphQLTag))),
      resolve: createProxyingResolver(async ({id}, _, {prisma: {tag}}) => {
        const tags = await tag.findMany({
          where: {
            authors: {
              some: {
                authorId: id
              }
            }
          }
        })

        return tags
      })
    }
  }
})

export const GraphQLAuthorFilter = new GraphQLInputObjectType({
  name: 'AuthorFilter',
  fields: {
    name: {type: GraphQLString},
    tagIds: {type: new GraphQLList(new GraphQLNonNull(GraphQLID))}
  }
})

export const GraphQLAuthorSort = new GraphQLEnumType({
  name: 'AuthorSort',
  values: {
    CREATED_AT: {value: AuthorSort.CreatedAt},
    MODIFIED_AT: {value: AuthorSort.ModifiedAt},
    NAME: {value: AuthorSort.Name}
  }
})

export const GraphQLAuthorConnection = new GraphQLObjectType<any, Context>({
  name: 'AuthorConnection',
  fields: {
    nodes: {type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(GraphQLAuthor)))},
    pageInfo: {type: new GraphQLNonNull(GraphQLPageInfo)},
    totalCount: {type: new GraphQLNonNull(GraphQLInt)}
  }
})

export const GraphQLAuthorLinkInput = new GraphQLInputObjectType({
  name: 'AuthorLinkInput',
  fields: {
    title: {type: new GraphQLNonNull(GraphQLString)},
    url: {type: new GraphQLNonNull(GraphQLString)}
  }
})

export const GraphQLAuthorInput = new GraphQLInputObjectType({
  name: 'AuthorInput',
  fields: {
    name: {type: new GraphQLNonNull(GraphQLString)},
    slug: {type: new GraphQLNonNull(GraphQLSlug)},
    links: {type: new GraphQLList(new GraphQLNonNull(GraphQLAuthorLinkInput))},
    bio: {type: GraphQLRichText},
    jobTitle: {type: GraphQLString},
    imageID: {type: GraphQLID},
    tagIds: {type: new GraphQLList(new GraphQLNonNull(GraphQLID))}
  }
})
