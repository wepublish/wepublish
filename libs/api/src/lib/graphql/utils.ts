import {Prisma} from '@prisma/client'
import {DateFilterComparison} from '../db/common'
import {AuthSession, AuthSessionType} from '@wepublish/authentication/api'

export const mapDateFilterToPrisma = (
  comparison: DateFilterComparison
): keyof Prisma.DateTimeFilter => {
  return comparison === DateFilterComparison.Equal ? 'equals' : comparison
}

export const isMeBySession = (id: string, session: AuthSession | null | undefined) =>
  session?.type === AuthSessionType.User && session.user.id === id
