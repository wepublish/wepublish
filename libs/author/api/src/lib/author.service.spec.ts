import {Test, TestingModule} from '@nestjs/testing'
import {AuthorService} from './author.service'
import {PrismaClient} from '@prisma/client'
import {AuthorDataloader} from './author.dataloader'
import {SortOrder} from '@wepublish/utils/api'
import {AuthorSort} from './author.model'

describe('AuthorService', () => {
  let service: AuthorService
  let dataloader: {[method in keyof AuthorDataloader]?: jest.Mock}
  let prismaMock: any

  beforeEach(async () => {
    prismaMock = {
      author: {
        findUnique: jest.fn(),
        findMany: jest.fn(),
        create: jest.fn(),
        update: jest.fn(),
        delete: jest.fn(),
        count: jest.fn()
      }
    }

    dataloader = {
      prime: jest.fn()
    }

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        AuthorService,
        {provide: AuthorDataloader, useValue: dataloader},
        {provide: PrismaClient, useValue: prismaMock}
      ]
    }).compile()

    service = module.get<AuthorService>(AuthorService)
  })

  it('should get an author by id', async () => {
    prismaMock.author.findUnique.mockResolvedValueOnce({
      id: '123',
      name: 'testAuthor',
      slug: 'test-author',
      links: []
    })
    await service.getAuthorById('123')
    expect(prismaMock.author.findUnique).toHaveBeenCalled()
    expect(prismaMock.author.findUnique.mock.calls[0]).toMatchSnapshot()
    expect(dataloader.prime).toHaveBeenCalled()
  })

  it('should get an author by slug', async () => {
    prismaMock.author.findUnique.mockResolvedValueOnce({
      id: '123',
      name: 'testAuthor',
      slug: 'test-author',
      links: []
    })
    await service.getAuthorBySlug('test-author')
    expect(prismaMock.author.findUnique).toHaveBeenCalled()
    expect(prismaMock.author.findUnique.mock.calls[0]).toMatchSnapshot()
    expect(dataloader.prime).toHaveBeenCalled()
  })

  it('should get authors with pagination', async () => {
    prismaMock.author.count.mockResolvedValueOnce(2)
    prismaMock.author.findMany.mockResolvedValueOnce([
      {
        id: '123',
        name: 'testAuthor',
        slug: 'test-author',
        links: []
      },
      {
        id: '234',
        name: 'anotherAuthor',
        slug: 'another-author',
        links: []
      }
    ])
    const args = {
      filter: {name: 'test'},
      sortedField: AuthorSort.Name,
      order: SortOrder.Ascending,
      skip: 0,
      take: 2
    }
    await service.getAuthors(args)
    expect(prismaMock.author.count).toHaveBeenCalled()
    expect(prismaMock.author.findMany).toHaveBeenCalled()
    expect(prismaMock.author.findMany.mock.calls[0]).toMatchSnapshot()
    expect(dataloader.prime).toHaveBeenCalled()
  })

  it('should create an author', async () => {
    const author = {
      name: 'testAuthor',
      slug: 'test-author',
      links: [],
      tagIds: []
    }
    prismaMock.author.create.mockResolvedValueOnce(author)
    await service.createAuthor(author)
    expect(prismaMock.author.create).toHaveBeenCalled()
    expect(prismaMock.author.create.mock.calls[0]).toMatchSnapshot()
  })

  it('should update an author', async () => {
    const author = {
      id: '123',
      name: 'updatedAuthor',
      slug: 'updated-author',
      links: [],
      tagIds: []
    }
    prismaMock.author.update.mockResolvedValueOnce(author)
    await service.updateAuthor(author)
    expect(prismaMock.author.update).toHaveBeenCalled()
    expect(prismaMock.author.update.mock.calls[0]).toMatchSnapshot()
    expect(dataloader.prime).toHaveBeenCalled()
  })

  it('should delete an author by id', async () => {
    prismaMock.author.delete.mockResolvedValueOnce({
      id: '123',
      name: 'deletedAuthor',
      slug: 'deleted-author',
      links: []
    })
    await service.deleteAuthorById('123')
    expect(prismaMock.author.delete).toHaveBeenCalled()
    expect(prismaMock.author.delete.mock.calls[0]).toMatchSnapshot()
    expect(dataloader.prime).not.toHaveBeenCalled()
  })
})
