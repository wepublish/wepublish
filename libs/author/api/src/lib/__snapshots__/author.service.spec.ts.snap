// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorService should create an author 1`] = `
[
  {
    "data": {
      "links": {
        "create": [],
      },
      "name": "testAuthor",
      "slug": "test-author",
      "tags": {
        "create": [],
      },
    },
    "include": {
      "links": true,
    },
  },
]
`;

exports[`AuthorService should delete an author by id 1`] = `
[
  {
    "include": {
      "links": true,
    },
    "where": {
      "id": "123",
    },
  },
]
`;

exports[`AuthorService should get an author by id 1`] = `
[
  {
    "include": {
      "links": true,
    },
    "where": {
      "id": "123",
    },
  },
]
`;

exports[`AuthorService should get an author by slug 1`] = `
[
  {
    "include": {
      "links": true,
    },
    "where": {
      "slug": "test-author",
    },
  },
]
`;

exports[`AuthorService should get authors with pagination 1`] = `
[
  {
    "cursor": undefined,
    "include": {
      "links": true,
    },
    "orderBy": {
      "name": "asc",
    },
    "skip": 0,
    "take": 3,
    "where": {
      "AND": [
        {
          "name": {
            "contains": "test",
            "mode": "insensitive",
          },
        },
        {},
      ],
    },
  },
]
`;

exports[`AuthorService should update an author 1`] = `
[
  {
    "data": {
      "links": {
        "create": [],
        "deleteMany": {
          "authorId": "123",
        },
      },
      "name": "updatedAuthor",
      "slug": "updated-author",
      "tags": {
        "connectOrCreate": [],
        "deleteMany": {
          "authorId": "123",
          "tagId": {
            "notIn": [],
          },
        },
      },
    },
    "include": {
      "links": true,
      "tags": true,
    },
    "where": {
      "id": "123",
    },
  },
]
`;
